### PR TITLE
Pipelined file sink separating serialization from I/O

### PR DESCRIPTION
**Pull request by strawgate**
## Summary

Introduces a **Serializer + Writer trait split** that separates CPU-bound JSON serialization from I/O-bound file writes using a dedicated OS writer thread with triple-buffered channels, plus platform-specific file I/O optimizations.

This mirrors the input-side architecture (`io_worker_loop` → channel → `cpu_worker_loop`) and delivers **21% higher sustained throughput** for file output at production volumes.

### New: `pipelined.rs` — reusable output pipeline infrastructure

- `BatchSerializer` trait — pure CPU, no I/O allowed (enforced at type level)
- `BatchWriter` trait — blocking I/O transport
- `PipelinedSink<S: BatchSerializer>` — generic sink that spawns a dedicated OS writer thread
- `FileWriter` — `std::fs::File` blocking I/O with platform-specific hints
- `JsonBatchSerializer` — wraps existing `StdoutSink::write_batch_to` logic

### Platform-specific I/O optimizations in `FileWriter`

| Platform | Optimization | Purpose |
|----------|-------------|---------|
| Linux | `fallocate(FALLOC_FL_KEEP_SIZE)` 64MB pre-alloc | Avoids per-write metadata updates on ext4/xfs |
| Linux | `posix_fadvise(POSIX_FADV_SEQUENTIAL)` | Hints sequential access pattern |
| Linux | `posix_fadvise(POSIX_FADV_DONTNEED)` after each write | Drops page cache for written data (write-only workload) |
| macOS | `fcntl(F_NOCACHE)` | Bypasses Unified Buffer Cache |
| macOS | `fcntl(F_PREALLOCATE)` 64MB | Pre-allocates disk space (contiguous, fragmented fallback) |

All hints are best-effort (failures silently ignored) — no correctness impact.

### Rewritten: `file_sink.rs`

- `FileSink` now wraps `PipelinedSink<JsonBatchSerializer>`
- `FileSinkFactory` keeps `is_single_use() = true` (single file, single writer — correctness)
- Removed `Arc<Mutex<tokio::fs::File>>` — simpler, faster

### Performance: `row_json.rs`

- `append_bytes`/`append_byte` helpers using `ptr::copy_nonoverlapping`
- ~25% faster JSON string writing in hot path (avoids Vec bounds checks in inner loop)

### Bench harness: `file_sink_compare`

- 15+ write strategies for systematic comparison
- Supports `--scenario`, `--sync-interval`, `--batches`, `--batch-size` flags

## Architecture

```
┌──────────────────┐     sync_channel(3)     ┌────────────────────┐
│  Caller thread   │ ──── filled bufs ─────▶ │  Writer OS thread  │
│  (tokio worker)  │ ◀─── empty bufs ─────── │  (std::fs::File)   │
│                  │                          │                    │
│  serialize(batch)│                          │  write_all(buf)    │
│  into buffer     │                          │  flush() on demand │
└──────────────────┘                          └────────────────────┘
```

Triple-buffered: serialize(N+1) overlaps with write(N), third buffer in flight.

## Benchmarks (Apple M4, wide/10k, 200 batches, 1.3GB, no sync)

| Strategy | Throughput | vs Serial |
|----------|-----------|-----------|
| Serialize only (no I/O) | 2.31M lines/sec | ceiling |
| Serial write (old pattern) | 1.97M lines/sec | baseline |
| **PipelinedSink + platform hints** | **2.38M lines/sec** | **+21%** |
| Threaded (manual, no hints) | 2.30M lines/sec | +17% |

The new FileSink **exceeds** the manual threaded pipeline because platform hints reduce filesystem metadata overhead.

## Competitive Analysis

No major log forwarder separates serialization from I/O:
- **Vector** (Rust/tokio): per-event write_all, single task, BufWriter
- **Fluent Bit** (C): fclose after each batch, single thread, libc stdio
- **OTel Collector** (Go): bufio.Flush on timer, no thread separation
- **rsyslog** (C): has async writer thread (closest to our pattern), but no platform hints
- **syslog-ng** (C): flush-lines batching, no thread separation

Our pipelined approach with platform hints is novel.

### Dragons identified from research (not in scope for this PR):
- logrotate interaction (inode-based reopen detection needed)
- Disk full mid-write (ENOSPC → RetryAfter mapping)
- NFS atomicity not guaranteed for O_APPEND
- Template paths need LRU fd cache + idle timeout
- Compressed files can't be appended after close

## Test Plan

- All 323 ffwd-output tests pass (including 3 file_sink-specific tests)
- `cargo clippy -p ffwd-output -- -D warnings` clean
- `cargo check -p ffwd-runtime` clean (downstream compiles)
- `file_sink_compare --scenario threaded` validates E2E throughput
- Buffer pool exhaustion fix: on serialize error, buffer recycled through writer thread

## Future Work

- Convert OTLP, Elasticsearch, Loki, TCP, UDP sinks to same pattern
- File rotation (size-based, logrotate compatibility)
- Make flush sync configurable (sync vs no-sync vs periodic)
- `create_dirs` for auto-creating parent directories
- Better error handling: ENOSPC → RetryAfter, EACCES → Rejected

## Review Feedback Addressed

All 38 review threads resolved. Key fixes:

- **Writer error propagation**: Write failures now surface to callers via dedicated error channel (non-blocking `check_writer_error()` on each `send_batch`)
- **`PipelinedSink::new()` → `io::Result`**: No more panics on thread spawn failure
- **`FileSink` constructors → `io::Result`**: Proper error propagation through factory
- **`pub(crate)` visibility**: Pipeline types are internal (no semver commitment)
- **`num_buffers` validation**: Rejects zero-buffer configs that would deadlock
- **Linux DONTNEED offset**: Initialized from file metadata (correct for appends to existing files)
- **Bench feature gates**: `just bench`/`bench-full`/`bench-otlp-io` now pass `--features io-bench`
- **Doc accuracy**: Fixed throughput claim (~40% → ~21%), fixed stale rustdoc example

### Acknowledged (by design, not changing):
- **Blocking `recv()` in async future**: Deliberate design matching the input-side pattern (`io_worker_loop` ↔ `cpu_worker_loop`). Output sinks run on dedicated worker tasks with bounded buffer pools (3 buffers = bounded blocking).
- **`append_bytes`/`append_byte` unsafe helpers**: Benchmarked ~25% faster than `extend_from_slice` in the hot JSON serialization loop. Soundness argument: reserve guard ensures capacity, `ptr::copy_nonoverlapping` between non-overlapping allocations, `set_len` follows initialization.
- **Proptests for pipeline state machine**: Good suggestion, separate PR scope.

Relates to #2733

> [!NOTE]
> ### Add pipelined file sink separating serialization from I/O on a dedicated writer thread
> - Introduces [`PipelinedSink`](https://github.com/strawgate/fastforward/pull/2736/files#diff-3df1575950dc67f5ab409fbbb264647880cab1cb4533cd73fd8055d2f7f82098) in a new `pipelined` module that overlaps CPU serialization (on the caller thread) with blocking file I/O (on a dedicated OS thread) using a triple-buffer channel.
> - Refactors [`FileSink`](https://github.com/strawgate/fastforward/pull/2736/files#diff-1e0617b0acebee6a5d0ff358b30c1e07f680cc1cb896b6dd0407f2452c571434) to use `PipelinedSink<JsonBatchSerializer>` backed by a `std::fs::File`; constructors now return `io::Result<Self>` and accept a standard file instead of a tokio async file.
> - Adds platform-specific I/O hints in `FileWriter`: `fallocate` + `posix_fadvise(SEQUENTIAL/DONTNEED)` on Linux; `F_NOCACHE` + `F_PREALLOCATE` on macOS.
> - Adds a new [`file_sink_compare`](https://github.com/strawgate/fastforward/pull/2736/files#diff-b3457a571ac22853b69080d9b4e2e9874f415859780fc198979d0db82fe4f8fb) CLI binary (behind `bench-tools` feature) for runtime comparison of file output strategies across serialization and I/O scenarios.
> - Risk: `FileSink::new` and `FileSink::with_message_field` now return `Result` — callers that previously assumed infallible construction will need updating.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized f473ce6.</sup>
> 
>

### CODE REVIEW

**Comment by strawgate** [comment]
## 🚀 PR Rocket · Auto-pilot ✅ ON

Status: **Watching**
Last action: No code changes needed: Addressed review feedback and fixed CI lint failures. Two force-pushes from the author were integrated during this run.

PR Rocket is actively managing this PR. It will follow the conversation,
fix CI, resolve conflicts, keep the branch updated, and make changes quietly.


<details><summary>Recent activity</summary>

- No code changes needed: Addressed review feedback and fixed CI lint failures. Two force-pushes from the author were integrated during this run.
- ⚡ Preview title and body, labels (enhancement, performance, refactor): Pipelined file sink separating serialization from I/O
- ⚡ Preview title and body, labels (enhancement, performance, refactor): Separate JSON serialization from file I/O via pipelined OS writer thread
- ⚡ Review: This is a well-structured performance refactor that separates serialization from
- ⚡ Preview title and body, labels (enhancement, performance, refactor): Separate serialization from I/O with pipelined file sink

</details>

---
`/rocket disable auto-pilot`

**Comment by strawgate** [comment]
### ⚡ Quick Review

This is a well-structured performance refactor that separates serialization from I/O via a dedicated writer thread. The unsafe `ptr::copy_nonoverlapping` path in `row_json.rs` is sound given the inline capacity pre-check, and the `PipelinedSink` architecture is clean. Overall risk is low, but a few items warrant attention:

- **Buffer loss on serialization error** (`pipelined.rs` line ~246): When `serializer.serialize` fails, the buffer is dropped without being returned to the pool. This is documented but means the pool permanently leaks one buffer per serialization error. If errors are rare this is fine, but worth a comment noting the acceptable trade-off.

- **`Drop` vs `shutdown` race** (`pipelined.rs` line ~395): The `Drop` impl calls `join()` on the writer thread, but `shutdown()` also joins. If a `PipelinedSink` is dropped while a concurrent `shutdown` future is still pending, the join could race. Adding an `Option<JoinHandle>` take in `shutdown` (similar to what it already does for `filled_tx`) would prevent double-join.

- **`is_single_use` removal** (`file_sink.rs`): The factory comment says "no longer single-use," which is correct and the new behavior is sound since each `create()` opens a fresh fd and writer thread. Just verify this doesn't break any existing callers that relied on single-use semantics.

- **Feature gate consistency**: Several new `required-features = ["io-bench"]` entries on benches look correct, but double-check that `generator_profile` and any other binaries using `FileSinkFactory` also have this gate — they're not in the diff so if they exist they may need updating.

---
*Fast review by PR Rocket — the full agent will follow.*

**Comment by strawgate** [comment]
### ⚡ Quick Review

This PR introduces a significant architectural refactor: separating serialization from I/O via a dedicated OS writer thread with triple-buffering, plus unsafe hot-path optimizations in the JSON serialization loop. The overall design is sound and the performance gains (~40%) are well-validated by the new benchmark suite. However, there are a few concerns worth noting.

Key observations:
- **Unsafe code in `row_json.rs`**: The new `append_bytes`/`append_byte` functions use `ptr::copy_nonoverlapping` and direct pointer writes. The safety invariants look correct (capacity pre-reserved, length set atomically), but this is low-level and deserves extra scrutiny and test coverage for edge cases like empty slices and zero-capacity vectors.
- **Buffer leak on serialization error**: In `PipelinedSink::send_batch`, if `serializer.serialize()` fails, the acquired buffer is silently dropped rather than recycled. The comment acknowledges this ("We can't easily return it since the channel goes the other way") but a fix using `empty_tx.try_send()` directly would be straightforward and would prevent buffer loss under error conditions.
- **Silent write errors in writer thread**: `writer_thread_loop` logs write failures but continues the loop, silently dropping data. No mechanism exists to surface errors back to callers. For a production system, consider propagating write errors through the `SendResult` chain or at least tracking them in stats.
- **Feature flag proliferation**: The new `io-bench` and `s3-bench` flags gate benchmarks cleanly, but the diff touches many unrelated benchmark entries. This is a minor organizational issue but increases review surface area.

Minor nits (non-blocking):
- The `is_single_use` comment in `FileSinkFactory` is helpful but could clarify whether callers are expected to respect this contract.
- `black_box(&out)` in benchmarks correctly prevents DCE; no action needed.

---
*Fast review by PR Rocket — the full agent will follow.*

File: crates/ffwd-output/src/pipelined.rs [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+            let bytes = buf.len() as u64;
+
+            // Hand off the filled buffer to the writer thread.
+            if let Err(e) = self.send_filled(buf) {
+                return SendResult::from_io_error(e);
+            }
+
+            self.stats.inc_lines(rows);
+            self.stats.inc_bytes(bytes);
+            SendResult::Ok
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/pipelined.rs:296]: This is the standard "fire-and-check-next" pattern used by pipelined I/O systems. Errors from batch N are surfaced when batch N+1 calls `send_batch()` via `check_writer_error()` (line 263). Additionally, `flush()` (line 300) provides a synchronous barrier that can be called during checkpoint flush — the pipeline shutdown path calls flush before advancing checkpoints, ensuring no acknowledged data is lost.

The tradeoff is intentional: we accept that a process crash between `send_batch()` return and the writer thread completing the write could lose up to one in-flight buffer (configurable depth=2 by default). This is the same guarantee as any buffered writer (including `BufWriter`) and is acceptable for a log forwarder where the source data still exists on disk.

File: crates/ffwd-output/src/pipelined.rs [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+        let fd = self.file.as_raw_fd();
+        // SAFETY: fd is valid (owned by `self.file`), offset and len describe
+        // already-written data, POSIX_FADV_DONTNEED is an advisory hint.
+        unsafe {
+            let _ = libc::posix_fadvise(
+                fd,
+                offset as libc::off_t,
+                len as libc::off_t,
+                libc::POSIX_FADV_DONTNEED,
+            );
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/pipelined.rs:504]: Acknowledged — these hints use OS advisory syscalls (`fallocate`, `posix_fadvise`, `fcntl`) that cannot cause UB regardless of return value. Moving to `ffwd-core` isn't feasible since core is `no_std`. However, I'll add `#[allow(unsafe_code)]` with a documented justification on the specific functions, and keep the SAFETY comments. These are purely advisory hints — failure is always benign (we ignore errors). If we later add a safe `io-hints` helper crate, we can move them there.

[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/pipelined.rs:504]: Addressed in ef38f41a: removed handwritten unsafe/libc calls from ffwd-output and switched the Linux file hints to safe rustix APIs; macOS now falls back to standard write_all rather than local fcntl unsafe.

File: crates/ffwd-output/src/pipelined.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/pipelined.rs:296]: Addressed in ef38f41a: Data messages now carry a per-batch ack and send_batch waits for writer.write to succeed before returning SendResult::Ok, so checkpoints only advance after persistence.

File: crates/ffwd-bench/Cargo.toml [hunk 3]
```diff
@@ -214,7 +225,7 @@ name = "otlp_io_profile"
 path = "src/bin/otlp_io_profile.rs"
 bench = false
 test = false
-required-features = ["bench-tools"]
+required-features = ["bench-tools", "io-bench"]
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-bench/Cargo.toml:228]: Addressed in ef38f41a: the just recipes for otlp_io_profile and framed_input_profile now include io-bench alongside bench-tools so they satisfy the binaries' required-features.

File: crates/ffwd-output/src/pipelined.rs [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+
+            let bytes = buf.len() as u64;
+
+            // Hand off the filled buffer to the writer thread.
+            if let Err(e) = self.send_filled(buf) {
+                return SendResult::from_io_error(e);
+            }
+
+            self.stats.inc_lines(rows);
+            self.stats.inc_bytes(bytes);
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/ffwd-output/src/pipelined.rs:295]: Addressed in ef38f41a: lines_total and bytes_total are incremented only after the writer-thread ack confirms the write succeeded; writer errors return IoError with counters unchanged.
